examples/conform.cpp

```diff
@@ -0,0 +1,163 @@
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+
+// Example OTIO script that reads a timeline and then relinks clips
+// to movie files found in a given folder, based on matching names.
+//
+// Demo:
+//
+// % ls -1R
+// editorial_cut.otio
+// media/
+//    shot1.mov
+//    shot17.mov
+//    shot99.mov
+//
+// % conform editorial_cut.otio media conformed.otio
+// Relinked 3 clips to new media.
+// Saved conformed.otio with 100 clips.
+//
+// % diff editorial_cut.otio conformed.otio
+// ...
+
+#include "util.h"
+
+#include <opentimelineio/clip.h>
+#include <opentimelineio/externalReference.h>
+#include <opentimelineio/timeline.h>
+
+#include <iostream>
+
+namespace otio = opentimelineio::OPENTIMELINEIO_VERSION;
+
+// Look for media with this name in this folder.
+std::string find_matching_media(std::string const& name, std::string const& folder)
+{
+    // In this case we're looking in the filesystem.
+    // In your case, you might want to look in your asset management system
+    // and you might want to use studio-specific metadata in the clip instead
+    // of just the clip name.
+    // Something like this:
+    // shot = asset_database->find_shot(
+    //    otio::any_cast<std::map<std::string, std::string> >(clip->metadata()["mystudio"])["shotID"]);
+    // new_media = shot->latest_render("mov");
+    
+    const auto matches = glob(folder, name + ".*");
+    
+    if (matches.size() == 0)
+    {
+        //std::cout << "DEBUG: No match for clip '" << name << "'" << std::endl;
+        return std::string();
+    }
+    if (matches.size() == 1)
+    {
+        return matches[0];
+    }
+    else
+    {
+        std::cout << "WARNING: " << matches.size() << " matches found for clip '" <<
+            name << "', using '" << matches[0] << "'";
+        return matches[0];
+    }
+}
+
+// Look for replacement media for each clip in the given timeline.
+//
+// The clips are relinked in place if media with a matching name is found.
+int conform_timeline(
+    otio::SerializableObject::Retainer<otio::Timeline> const& timeline,
```

Comment:
conform_timeline doesn't do anything that will change the retain count on the Timeline, so passing a Retainer isn't required in this case. Passing Timeline const& might emphasize that point.

OTOH, the Retainer semantics are difficult to build a mental model for, so it might be reasonable in a tutorial to do the unbreakable thing.

Would you mind adding some notes around why a Retainer is used here, to help people using this code in copy-pasta later?

Reply:
I guess I was thinking of Retainer like std::shared_ptr, and that it would be a good example to always use it to point to OTIO objects? I you don't think that's a good idea I'm happy to change it to a regular pointer.

Reply:
I'm happy to have just a line of comment there as you'd find in a tutorial, just to explain that Retainer manages the internal reference count on the retained object. (It's an intrusive pointer system, rather than a shared pointer).